torch/distributed/optim/functional_adam.py

```diff
@@ -138,6 +141,7 @@ def step(self, gradients: List[Optional[Tensor]]):
 
         for param, gradient in zip(self.param_group["params"], gradients):
             if gradient is not None:
+                has_complex |= torch.is_complex(param)
```

Comment:
This should likely be 
```suggestion
                has_complex |= torch.is_complex(param) || torch.is_complex(gradient)
```
Though I'm curious, @albanD @lezcano is there a known use case where the params and grads differ in complexness?

Reply:
For the stateful version, it is technically impossible for the .grad field to be of different dtype from the Tensor.
You can bypass that with the functional call. But I think it's ok to assume same complex-ness here.

Reply:
Right, I guess we should be worried about functional use-case. 

I think its quite unlikely though, and people will get errors if they try to do this now, so at least it wouldn't go unnnoticed.

Reply:
Anw, my justification for not needing this (which I believe is quite sound): https://github.com/pytorch/pytorch/pull/110613#discussion_r1347853161

In the single tensor case, we only every check `param`

Reply:
Ah, then yea, heuristically this is fine to not have. The other comment I put in the other PR about checking is_complex on each state for the functional APIs is still valid though.

Reply:
> checking is_complex on each state for the functional APIs is still valid though.

Hmm could you check again, I'm trying to say that the single_tensor case never does this check, so multi_tensor case should not either.